### PR TITLE
add server to filtered cipher suites log

### PR DESCRIPTION
**Pull request by ramaraochavali**
[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[X ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[X] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

### CODE REVIEW

**Comment by ramaraochavali** [comment]
@howardjohn would this work?

File: pilot/pkg/model/gateway.go [hunk 1]
```diff
@@ -167,6 +168,13 @@ func MergeGateways(gateways []gatewayWithInstances) *MergedGateway {
 					if s.Tls.Mode == networking.ServerTLSSettings_AUTO_PASSTHROUGH {
 						autoPassthrough = true
 					}
+					if len(s.Tls.CipherSuites) > 0 {
+						serverName := s.Name
+						if len(serverName) == 0 {
+							serverName = s.Port.Name
```
Comment (hzxuzhonghu) [inline pilot/pkg/model/gateway.go:174]: This is hacky

**Review by hzxuzhonghu — COMMENTED** [review]
lgtm

**Comment by hzxuzhonghu** [comment]
Just noticed the issue. For the previous gateway with invalid cipher suites, it is invalid with nack. I would rather prevent it in validation. We can assume no such gateways exist in mesh.

File: pilot/pkg/model/gateway.go (hunk 1 above)
Comment (ramaraochavali) [reply pilot/pkg/model/gateway.go:174]: So we either identify server with unique port name or server name. Why do you think it is hacky?

**Comment by ramaraochavali** [comment]
https://github.com/istio/istio/pull/34109 - We discussed about that here and decided to not prevent such gateways but add warning because a gateway may have many servers and only a single cipher suite may be invalid in server.

File: pilot/pkg/model/gateway.go [hunk 2]
```diff
@@ -167,6 +168,13 @@ func MergeGateways(gateways []gatewayWithInstances) *MergedGateway {
 					if s.Tls.Mode == networking.ServerTLSSettings_AUTO_PASSTHROUGH {
 						autoPassthrough = true
 					}
+					if len(s.Tls.CipherSuites) > 0 {
+						serverIdentifier := s.Name
+						if len(serverIdentifier) == 0 {
+							serverIdentifier = s.Port.Name
+						}
+						s.Tls.CipherSuites = filteredCipherSuites(s.Tls.CipherSuites, serverIdentifier, gatewayName)
```
Comment (howardjohn) [inline pilot/pkg/model/gateway.go:176]: is it safe to mutate here? Will this impact the underlying client-go cache?

Comment (hzxuzhonghu) [reply pilot/pkg/model/gateway.go:176]: I didnot realize it is mutating the original object

Comment (ramaraochavali) [reply pilot/pkg/model/gateway.go:176]: So is it OK if filteredCipherSuite returns a different set, we clone the server and set?

Comment (hzxuzhonghu) [reply pilot/pkg/model/gateway.go:176]: What is this pr fixing?

Comment (ramaraochavali) [reply pilot/pkg/model/gateway.go:176]: #34109 added validation and logic to skip invalid cipher suites. But the logic to skip is executed on every listener creation. I am trying to see if we can do it once with this PR.

**Comment by ramaraochavali** [comment]
I tried to clone when ciphersuites are invalid - but selective cloning does not work as the same server is referenced in multiple places like serversByRouteName. I felt cloning every server is not efficient - rather we pay the cost of comparison at listener creation. So I reverted that change and just added server to log to indicate which server is causing issue.

File: pilot/pkg/networking/core/v1alpha3/gateway.go [hunk 3]
```diff
@@ -992,3 +979,17 @@ func buildGatewayVirtualHostDomains(hostname string, port int) []string {
...
+
+// Invalid cipher suites lead Envoy to NACKing. This filters the list down to just the supported set.
+func filteredCipherSuites(server *networking.Server) []string {
+	suites := server.Tls.CipherSuites
+	ret := make([]string, 0, len(suites))
+	for _, s := range suites {
+		if security.IsValidCipherSuite(s) {
+			ret = append(ret, s)
+		} else {
+			log.Debugf("ignoring unsupported cipherSuite: %q for server %s", s, server.String())
```
Comment (howardjohn) [inline pilot/pkg/networking/core/v1alpha3/gateway.go:991]: nit: server.String is fairly expensive. Maybe we can wrap in `if log.DebugEnabled` so we do not pay for it when debug log is off

**Review by hzxuzhonghu — APPROVED** [review]

**Pull request merged**